src/action.py

```diff
@@ -140,39 +137,55 @@ def _update(
         raise exceptions.ActionError(
             f"internal error, content change for page is None, {action=!r}"
         )
-    if (
-        action.navlink_change.new.link is not None
-        and action.content_change is not None
-        and action.content_change.new is None
-    ):
-        raise exceptions.ActionError(f"internal error, new content for page is None, {action=!r}")
 
     if (
         action.content_change is not None
         and action.content_change.old is not None
         and action.content_change.new is not None
     ):
-        _log_content_change(old=action.content_change.old, new=action.content_change.new)
+        _log_content_change(
+            base=action.content_change.base or action.content_change.old,
+            new=action.content_change.new,
+        )
 
-    if (
+    # All of these are needed to ensure types are as expected
+    if (  # pylint: disable=too-many-boolean-expressions
         not dry_run
         and action.navlink_change.new.link is not None
         and action.content_change is not None
+        and action.content_change.base is not None
         and action.content_change.new is not None
         and action.content_change.new != action.content_change.old
     ):
         try:
-            discourse.update_topic(
-                url=action.navlink_change.new.link, content=action.content_change.new
+            merged_content = content_merge(
+                base=action.content_change.base,
+                theirs=action.content_change.old,
+                ours=action.content_change.new,
             )
+            discourse.update_topic(url=action.navlink_change.new.link, content=merged_content)
             result = types_.ActionResult.SUCCESS
             reason = None
         except exceptions.DiscourseError as exc:
             result = types_.ActionResult.FAIL
             reason = str(exc)
+        except exceptions.ContentError as exc:
+            result = types_.ActionResult.FAIL
+            reason = str(exc)
     else:
-        result = types_.ActionResult.SKIP if dry_run else types_.ActionResult.SUCCESS
-        reason = DRY_RUN_REASON if dry_run else None
+        if dry_run:
+            result = types_.ActionResult.SKIP
+            reason = DRY_RUN_REASON
+        elif (
+            action.content_change is not None
+            and action.content_change.base is None
+            and action.content_change.new != action.content_change.old
```

Comment:
Yes correct, the `action` module doesn't know for sure that it won't be called with conflicts so it is better to handle the exception rather than let it fail in a mess